### PR TITLE
Hierarchies: Grouping refactor

### PR DESCRIPTION
**Pull request by grigasp**
Closes https://github.com/iTwin/presentation/issues/434

### CODE REVIEW

File: packages/hierarchies/src/hierarchies/internal/operators/Grouping.ts [hunk 1]
```diff
@@ -88,34 +100,33 @@ async function groupInstanceNodes(
   nodes: ProcessedInstanceHierarchyNode[],
   extraSiblings: number,
   groupingHandlers: GroupingHandler[],
-  onGroupingNodeCreated: undefined | ((groupingNode: ProcessedGroupingHierarchyNode) => void),
+  parentNode: ParentHierarchyNode | undefined,
```
Comment (Yato333) [inline packages/hierarchies/src/hierarchies/internal/operators/Grouping.ts:103]:
```suggestion
  parentNode?: ParentHierarchyNode,
```

File: packages/hierarchies/src/hierarchies/internal/operators/Grouping.ts [hunk 2]
```diff
@@ -88,34 +100,33 @@ async function groupInstanceNodes(
...
-            }),
-          ),
-        ),
-        compareNodesByLabel,
-      ),
+      grouped: [...(curr?.grouped ?? []), ...groupings.grouped],
       ungrouped: groupings.ungrouped,
     };
   }
   if (curr) {
```
Comment (Yato333) [inline packages/hierarchies/src/hierarchies/internal/operators/Grouping.ts:115]: We can reduce the amount of nesting here. How about this:
```ts
  if (!curr) {
    return nodes;
  }

  for (const groupingNode of curr.grouped) {
    onGroupingNodeCreated?.(groupingNode);
    if (!parentNode) {
      continue;
    }

    if (HierarchyNode.isGroupingNode(parentNode)) {
      groupingNode.nonGroupingAncestor = parentNode.nonGroupingAncestor;
      continue;
    }

    // not sure why type checker doesn't pick this up
    assert(HierarchyNode.isCustom(parentNode) || HierarchyNode.isInstancesNode(parentNode));
    groupingNode.nonGroupingAncestor = parentNode;
  }

  return [...curr.grouped, ...curr.ungrouped];
```

File: packages/hierarchies/src/hierarchies/internal/operators/Grouping.ts (hunk 1 above)
Comment (grigasp) [reply packages/hierarchies/src/hierarchies/internal/operators/Grouping.ts:103]: In this context parent node is not optional - `undefined` means we're requesting root nodes. So I'd prefer to leave it as is.

File: packages/hierarchies/src/hierarchies/internal/operators/grouping/PropertiesGrouping.ts [hunk 3]
```diff
@@ -44,18 +46,45 @@ export interface PropertyGroupInfo {
...
+            properties: [],
+          },
+          parentKeys: [],
+          groupedInstanceKeys: [],
+          label: localizedStrings.other,
+          children: [],
+        },
+        new: true,
+      };
+    }
```
[the remark this answers is not in the document]
Comment (grigasp) [reply packages/hierarchies/src/hierarchies/internal/operators/grouping/PropertiesGrouping.ts]: applied

File: packages/hierarchies/src/hierarchies/internal/operators/Grouping.ts (hunk 2 above)
Comment (grigasp) [reply packages/hierarchies/src/hierarchies/internal/operators/Grouping.ts]: agree, fixed

File: packages/hierarchies/src/hierarchies/internal/operators/grouping/LabelGrouping.ts [hunk 4]
```diff
@@ -57,8 +56,8 @@ export async function createLabelGroups(nodes: ProcessedInstanceHierarchyNode[])
   });
 
   return {
-    grouped: sortNodesByLabel(groupedNodes),
-    ungrouped: mergeSortedArrays(ungrouped, sortNodesByLabel(mergedNodes), compareNodesByLabel),
+    grouped: groupedNodes,
+    ungrouped: [...ungrouped, ...mergedNodes],
     groupingType: "label",
   };
```
[the remark this answers is not in the document]
Comment (grigasp) [reply packages/hierarchies/src/hierarchies/internal/operators/grouping/LabelGrouping.ts]: applied suggestion here and in some other places too

**Review by Yato333 — APPROVED** [review]

**Pull request merged**